udev/net_id: don't generate PCI hotplug slot based names if two PCI devices claim the same slot

msekletar opened the pull request:
Before commit 9009d3b5c3b6d191be69215736be77583e0f23f9 we would try to
match PCI device address with a slot index only using parent device of
a network device. Now, we walk the entire chain of PCI devices because
sometimes the slot is associated with a PCI device further up than the
parent.

This creates a problem in case when there are two or more network
devices chained off of external PCI bridge. In such case parent device
(bridge) is associated with the slot and slot based naming would
produce the same name for all child network devices.

To address the issue we detect this situation for each network device.
We examine all PCI device chains of all network devices and try to
figure out whether there is some other PCI device that would claim the
same index as the device that we are handling now. If so, we refuse to
generate slot based name and hence we automatically fallback to some
other naming scheme.

msekletar commented:
@stuarthayes it would be much appreciated if you could review this.

poettering commented:
hmm, this should probably get a new naming policy flag?

yuwata commented:
CI failed.
```
../src/udev/udev-builtin-net_id.c:365:48: error: unused variable 'pci' [-Werror,-Wunused-variable]
        _cleanup_(sd_device_unrefp) sd_device *pci = NULL;
                                               ^
../src/udev/udev-builtin-net_id.c:366:33: error: unused variable 'dir' [-Werror,-Wunused-variable]
        _cleanup_closedir_ DIR *dir = NULL;
                                ^
2 errors generated.
```

msekletar commented:
> hmm, this should probably get a new naming policy flag?

I don't think so. This PR doesn't introduce new policy or change to existing one. We just skip generating PCI index base name in case there is a conflict and two or more devices would try claim the same name.

michich commented:
Isn't enumerating other devices racy? What if the other net devices under the same PCI bridge have not been yet probed by the kernel?

msekletar commented:
> Isn't enumerating other devices racy? What if the other net devices under the same PCI bridge have not been yet probed by the kernel?

You are correct, that is the problem. However, current code already has that problem. Right now it undefined which device will win the race and claim conflicting name. This patch tries to avoid the problem if possible and otherwise will leave the situation as is. Honestly, I don't know if there is a way around this except reverting the commit that introduced the issue. However, that would most certainly break network naming on some Dell servers.

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
+        if (r < 0)
+                return r;
+
+        slots = path_join(syspath, slots);
```
missing double quotations: `path_join(syspath, "slots");`

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
+
+                        str = path_join(slots, dent->d_name, "address");
+                        /* match slot address with device by stripping the function */
+                        if (str && read_one_line_file(str, &address) >= 0 && startswith(sysname, address)) {
```
I'd like to return `-ENOMEM` here on oom.

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
+                                break;
+
+                        if (!pcidev)
+                                break;
```
This condition seems not necessary, i.e., if `r >= 0`, `pcidev` is always set.

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
+        /* Iterate over all network devices and detect whether there is some other network
+           device that would want to claim same PCI hotplug slot index */
+        FOREACH_DEVICE(e, d) {
+                sd_device *pcidev = NULL;
```
So, it seems not necessary to initialize `pcidev` here.

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -305,54 +421,10 @@ static int dev_pci_slot(sd_device *dev, struct netnames *names) {
+        /* figure out PCI hotplug index, but use it as part of the name only if
+           there is no other device claiming the same index */
+        hotplug_slot = pci_hotplug_slot_index(names->pcidev);
+        if (hotplug_slot > 0 && !pci_hotplug_slot_conflict(names->pcidev, hotplug_slot)) {
```
Maybe we should propagate negative errno by `pci_hotplug_slot_index()` and `_conflict()`.

msekletar commented:
I think that doing conflict detection later in `link_config_apply()` would not really help to prevent the race pointed out by @michich. Also `link_config_apply()` currently is concerned with setting up actual network related configuration, moving HW specific code there feels like a layering violation.

msekletar commented:
@yuwata I've addressed your review comments and force-pushed my branch.

Hopefully, I didn't break anything while trying to fix things up. I don't have a test system available at the moment unfortunately so changes are only compile-tested.

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
+                const char *sysname;
+
+                if (sd_device_get_sysname(dev, &sysname) < 0)
+                        continue;
```
If this fails, the loop become infinite... The error should be propagate.

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
+
+        /* Iterate over all network devices and detect whether there is some other network
+           device that would want to claim same PCI hotplug slot index */
+        FOREACH_DEVICE(e, d) {
```
`{` is not necessary.

yuwata commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -305,54 +421,19 @@ static int dev_pci_slot(sd_device *dev, struct netnames *names) {
-        r = sd_device_get_syspath(pci, &syspath);
+        hotplug_slot = r;
+
+        r = pci_hotplug_slot_conflict(names->pcidev, hotplug_slot);
```
`_conflict()` is not necessary to be called if `hotplug_slot == 0`.

yuwata commented:
> I think that doing conflict detection later in `link_config_apply()` would not really help to prevent the race pointed out by @michich.

Right.

> Also `link_config_apply()` currently is concerned with setting up actual network related configuration, moving HW specific code there feels like a layering violation.

It makes some sense. But for such a HW, I think we cannot solve the conflict and the race completely. So, I think we need to warn about the problem, and encourage users to remove `slot` from the corresponding .link files. And I think the right place to logs about that is `link_config_apply()`.

poettering commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
+                        break;
+
+                rewinddir(dir);
+        }
```
I agree with @michich: this cannot work if slots are probed concurrently with this code. if slots appear in sysfs as our code here runs this will go wrong.

I am not convinced this is the right way to go if it's not reliable.

poettering commented:
Instead of this, can#t we just stop iterating up the tree as soon as we hit a PCI bridge? i.e. instead of trying to find siblings of PCI slots (which is racy afaiu) just stop when there could be siblings and not do our slot magic at all?

msekletar commented:
Closed in favour of #17352
